Raven db 13655

talweiss1982 opened the pull request.

Commit:
RavenDB-13655 skipping test that throws SO on 32-bit and avoiding AVE on cluster tx one by one

Commit:
RavenDB-13655 removing nullification of DocumentDatabase in DisposeIndex since it is causing race conditions during dispose and can cause NREs

Commit:
RavenDB-13655
* making sure database is deleted in test base
* nullifying conflict resolver in replication handler since it came up in gcroot for dead DocumentDatabase (from finilization queue

Commit:
RavenDB-13655 skipping IndexingWhenEncryptedTransactionSizeLimitLimitExceeded since somebody changed the indexing transaction size

Commit:
RavenDB-13655 fixing subscription test that seems to fail on 32-bit

Commit:
RavenDB-13655 fixing timeout when deleting database

ayende commented on test/SlowTests/Issues/RavenDB_12151.cs:
```diff
@@ -116,7 +116,7 @@ public void IndexingWhenGlobalScratchSpaceLimitExceeded32()
         }
 
-        [Fact]
+        [Fact(Skip ="Transaction size changed below 1MB")]
```
Why this change?

ayende replied:
What changed the TX size? We need to fix the test / figure out why it changed

talweiss1982 replied:
hard to track what caused it, @maximburyak and I decided to finish fixing the rest of the tests and try to figure this one on a later PR since it was taking too long and I couldn't track what caused it.
I know for sure this test was passing when I wrote it.
now the largest Tx is 334kb...

talweiss1982 replied:
It is likely the limitations I have put on the merger that are limiting the indexing transactions. I'll open a separate issue to investigate this but for the short term let's get the tests green again.

talweiss1982 replied:
It is likely the limitations I have put on the merger that are limiting the indexing transactions. I'll open a separate issue to investigate this but for the short term let's get the tests green again.

maximburyak replied:
That was resolved in the commit below

talweiss1982 committed:
RavenDB-13655 making IndexingWhenEncryptedTransactionSizeLimitLimitExceeded pass again
Also added consideration for read transaction size when running in encrypted environment

talweiss1982 committed:
RavenDB-13655 catching operation canceled exception during the deletion of a database in RavenTestBase

maximburyak commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -3278,6 +3276,9 @@ private int MinBatchSize
             var txAllocationsInBytes = UpdateThreadAllocations(indexingContext, indexWriteOperation, stats, updateReduceStats: false);
 
+            //We need to take the read transaction encryption size into account as we might read alot of document and produce very little indexing output.
+            txAllocationsInBytes += documentsOperationContext.Transaction.InnerTransaction.LowLevelTransaction.TotalEncryptionBufferSize.GetValue(SizeUnit.Bytes);
```
Nice

maximburyak commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -1217,7 +1217,9 @@ public void Dispose()
             });
 
-            ea.Execute(() => ConflictResolver?.ResolveConflictsTask.Wait());
+            ea.Execute(() => ConflictResolver?.ResolveConflictsTask.Wait(TimeSpan.FromSeconds(15)));
```
I'm not sure about the consequences of not waiting through here, and we do a similar unbounded wait few line above, although I agree that unbounded waits in dispose function is a bad practice in general.
```suggestion
            ea.Execute(() => ConflictResolver?.ResolveConflictsTask.Wait());
```

maximburyak commented on test/SlowTests/Client/Subscriptions/RavenDB-8404.cs:
```diff
@@ -48,7 +48,11 @@ public async Task SubscriptionsRQLSupportStartsWith()
                     users.AddRange(x.Items.Select(i => i.Result));
-                    amre.Set();
+                    //in 32bit we seem to send batches of 1 User
+                    if (users.Count == 3)
```
Right, because we started limiting the batch size in subscriptions

Commit:
Update src/Raven.Server/Documents/Replication/ReplicationLoader.cs

Co-Authored-By: maximburyak <maxim@ayende.com>

ayende reviewed (approved):
Very nice work.
